### PR TITLE
Functionality to capture user activities

### PR DESCRIPTION
**Pull request by Sachinbisht27**
#17 

### Types of changes

- [ ] Bug fix (a change which fixes an issue)
- [x] New feature (a change which adds functionality)

### Short description of what this resolves:
**Changes contain:** 
- Handling of webhook to get user activity details
- Addition of the logic to create a new entry in the user activity table.
- Logic to update the details in the same entry once the flow starts, is successful and is complete.

**Testing payloads:**
- For activity started:
```JSON
{
  "contact": {
    "id": 15356,
    "name": "User Name",
    "fields": {
      "dost_content_completed": {
        "inserted_at": "2024-04-19T11:44:41.961644Z",
        "label": "dost_content_completed",
        "type": "string",
        "value": "activity_1_sent"
      },
      "flow_execution_date": {
        "inserted_at": "2023-03-22T07:52:03.097151Z",
        "label": "flow_execution_date",
        "type": "string",
        "value": "Invalid Code"
      },
      "activity_1_started": {
        "label": "activity_1_started",
        "type": "string",
        "value": "true",
        "inserted_at": "2024-04-29T04:32:45.688102Z"
      },
      "last_flow": {
        "inserted_at": "2023-03-22T07:40:27.888117Z",
        "label": "last_flow",
        "type": "string",
        "value": "AWS_1"
      }
    },
    "phone": "+911234567890"
  },
  "current_time": "@calendar",
  "custom_key": "custom_value",
  "flow_status": "started",
  "flow_id": "14194",
  "flow_name": "Glific_Integration_Test_Flow",
  "flow_type": "activity",
  "flow_uuid": "3dddce1e-b76d-42d8-834b-964d70e3a0f9",
  "organization_id": 11,
  "activity_1_started": "yes",
  "results": {
    "flow_keyword": {
      "category": "draft:whatsappinit",
      "input": "draftwhatsappinit",
      "inserted_at": "2024-04-29T04:32:45.610016Z"
    }
  }
}
```
-  For activity suceeded:
```JSON
{
  "contact": {
    "id": 15356,
    "name": "User Name",
    "fields": {
      "dost_content_completed": {
        "inserted_at": "2024-04-19T11:44:41.961644Z",
        "label": "dost_content_completed",
        "type": "string",
        "value": "activity_1_sent"
      },
      "flow_execution_date": {
        "inserted_at": "2023-03-22T07:52:03.097151Z",
        "label": "flow_execution_date",
        "type": "string",
        "value": "Invalid Code"
      },
      "activity_1_success": {
        "label": "activity_1_success",
        "type": "string",
        "value": "true",
        "inserted_at": "2024-04-29T04:32:45.688102Z"
      },
      "last_flow": {
        "inserted_at": "2023-03-22T07:40:27.888117Z",
        "label": "last_flow",
        "type": "string",
        "value": "AWS_1"
      }
    },
    "phone": "+911234567890"
  },
  "current_time": "@calendar",
  "custom_key": "custom_value",
  "flow_status": "started",
  "flow_id": "14194",
  "flow_name": "Glific_Integration_Test_Flow",
  "flow_type": "activity",
  "flow_uuid": "3dddce1e-b76d-42d8-834b-964d70e3a0f9",
  "organization_id": 11,
  "activity_1_success": "yes",
  "results": {
    "flow_keyword": {
      "category": "draft:whatsappinit",
      "input": "draftwhatsappinit",
      "inserted_at": "2024-04-29T04:32:45.610016Z"
    }
  }
}
```
-  For activity completed:
```JSON
{
  "contact": {
    "id": 15356,
    "name": "User Name",
    "fields": {
      "dost_content_completed": {
        "inserted_at": "2024-04-19T11:44:41.961644Z",
        "label": "dost_content_completed",
        "type": "string",
        "value": "activity_1_sent"
      },
      "flow_execution_date": {
        "inserted_at": "2023-03-22T07:52:03.097151Z",
        "label": "flow_execution_date",
        "type": "string",
        "value": "Invalid Code"
      },
      "activity_1_completed": {
        "label": "activity_1_completed",
        "type": "string",
        "value": "true",
        "inserted_at": "2024-04-29T04:32:45.688102Z"
      },
      "last_flow": {
        "inserted_at": "2023-03-22T07:40:27.888117Z",
        "label": "last_flow",
        "type": "string",
        "value": "AWS_1"
      }
    },
    "phone": "+911234567890"
  },
  "current_time": "@calendar",
  "custom_key": "custom_value",
  "flow_id": "14194",
  "flow_name": "Glific_Integration_Test_Flow",
  "flow_type": "activity",
  "flow_status": "completed",
  "flow_uuid": "3dddce1e-b76d-42d8-834b-964d70e3a0f9",
  "organization_id": 11,
  "activity_1_completed": "yes",
  "results": {
    "flow_keyword": {
      "category": "draft:whatsappinit",
      "input": "draftwhatsappinit",
      "inserted_at": "2024-04-29T04:32:45.610016Z"
    }
  }
}
```

### Checklist:

- [x] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [x] I have documented my code wherever required.
- [ ] The changes requires a change to the documentation.
- [ ] I have updated the documentation based on the my changes.
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.\

### CODE REVIEW

File: api/models/user_activities.py [hunk 1]
```diff
@@ -1,18 +1,49 @@
+from datetime import date
+from flask_sqlalchemy.query import Query as BaseQuery
+from sqlalchemy import desc, func
+
 from api import db
 from api.mixins import TimestampMixin
 
 
+class UserActivitiesQuery(BaseQuery):
+    def get_started_activity_for_user(self, user_id, user_phone):
```
Comment (Satendra-SR) [inline api/models/user_activities.py:10]: shouldn't it be like this?
```suggestion
    def get_todays_started_activity_for_user(self, user_id, user_phone):
```

File: api/models/user_activities.py [hunk 2]
```diff
@@ -1,18 +1,49 @@
...
+                UserActivities.user_id == user_id,
+                UserActivities.user_phone == user_phone,
+                UserActivities.is_started is True,
+                func.DATE(UserActivities.started_on) == date.today(),
+            )
+            .order_by(desc("started_on"))
+            .first()
+        )
+
+    def get_succeeded_activity_for_user(self, user_id, user_phone):
```
Comment (Satendra-SR) [inline api/models/user_activities.py:22]: same here

File: api/models/user_flows.py [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
 
 
 class UserFlowsQuery(BaseQuery):
-    def get_by_flow_uuid_and_phone(self, flow_uuid, user_phone):
-        return self.filter(
-            UserFlows.flow_uuid == flow_uuid,
-            UserFlows.user_phone == user_phone,
-            func.DATE(UserFlows.flow_start_time) == date.today(),
-        ).first()
+    def get_latest_flow_log_today(self, flow_uuid, user_phone):
```
Comment (Satendra-SR) [inline api/models/user_flows.py:10]:
```suggestion
    def get_todays_latest_user_flow(self, flow_uuid, user_phone):
```

File: api/services/flow_run_log_service.py [hunk 4]
```diff
@@ -15,18 +14,18 @@ def create_user_flow_log(self, json_data):
...
                 flow_uuid, self.user.phone
             )
 
             user_flow_log = None
 
-            if not today_flow_log:
+            if flow_status == self.class_model.FlowRunStatus.COMPLETED:
+                user_flow_log = self.update_log(latest_flow_log)
+            else:
                 user_flow_log = self.create_log(flow_uuid, flow_name, flow_type)
```
Comment (Satendra-SR) [inline api/services/flow_run_log_service.py:28]: We should create a new log only if the start node is available - ie flow_status == "Started", thoughts?

File: api/services/user_activities_service.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class UserActivitiesService:
+    def __init__(self, user, user_flow):
+        self.user_id = user.id
+        self.user_phone = user.phone
+        self.user_flow_id = user_flow.id
+        self.class_model = models.UserActivities
+
+    def handle_user_activities(self, json_data: Dict[str, Any]):
```
Comment (Satendra-SR) [inline api/services/user_activities_service.py:16]: Can we make it consistent? `json_data: Dict[str, Any]`?

File: api/services/user_activities_service.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+            db.session.commit()
+            logger.info(f"Captured user activity for {self.user_phone}.")
+        except Exception as e:
+            logger.error(
+                f"Error while capturing user activity for {self.user_phone}."
+                f"Error message: {e}"
+            )
+
+    def create_or_update_user_activity(
```
Comment (Satendra-SR) [inline api/services/user_activities_service.py:37]: Let's divide this function into multiple sub-functions that will perform only "Single Responsibility"?

**Review by Satendra-SR — CHANGES_REQUESTED** [review]
@Sachinbisht27 a few changes are required. Have a look

File: api/services/flow_run_log_service.py (hunk 4 above)
Comment (Sachinbisht27) [reply api/services/flow_run_log_service.py:28]: Yes, thanks for the suggestion. Let me update that.

File: api/services/user_activities_service.py (hunk 6 above)
Comment (Sachinbisht27) [reply api/services/user_activities_service.py:37]: Sure @Satendra-SR. Let me do this.

File: api/services/user_activities_service.py (hunk 5 above)
Comment (Sachinbisht27) [reply api/services/user_activities_service.py:16]: Sure

Comment (Sachinbisht27) [reply api/services/user_activities_service.py:16]: Sure we can do this. Thanks for sharing.

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -30,3 +30,11 @@ repos:
   rev: v1.7.5
   hooks:
     - id: docformatter
+-   repo: https://github.com/pre-commit/pygrep-hooks
+    rev: v1.10.0
+    hooks:
+    -   id: python-use-type-annotations
+- repo: https://github.com/sondrelg/pep585-upgrade
```
Comment (Satendra-SR) [inline .pre-commit-config.yaml:37]: Why do we need this? 
The GitHub repo is archived now, if this is helpful , we may need to find an alternative -  https://github.com/sondrelg/pep585-upgrade

File: api/services/user_activities_service.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        if is_started:
+            return self.create_started_activity(activity_key, current_ist_time)
+        elif is_succeeded:
+            return self.update_succeeded_activity(current_ist_time)
+        elif is_completed:
+            return self.update_completed_activity(current_ist_time)
+        else:
+            return None
+
+    def create_started_activity(
```
Comment (Satendra-SR) [inline api/services/user_activities_service.py:59]: create Start activity should be similar to create activity, no? Can we do it like this in that case?
```suggestion
    def create_activity(
```

File: api/services/user_activities_service.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            user_id=self.user_id,
+            user_phone=self.user_phone,
+            user_flow_id=self.user_flow_id,
+            activity=activity_key.strip(),
+            is_started=True,
+            started_on=current_ist_time,
+            is_succeeded=None,
+            succeeded_on=None,
+            is_completed=None,
+            completed_on=None,
```
Comment (Satendra-SR) [inline api/services/user_activities_service.py:72]: don't we have default column values as `None` for these columns? What if we don't add them from here?

**Review by Satendra-SR — COMMENTED** [review]
@Sachinbisht27 a few comments before we merge the changes. Have a look

File: .pre-commit-config.yaml (hunk 7 above)
Comment (Sachinbisht27) [reply .pre-commit-config.yaml:37]: I was exploring some pre-commit hooks so, we can maintain the TypeHinting in the codebase. I found 2 hooks related to it so added them here. I can explore more if this one is archived now.

File: api/services/user_activities_service.py (hunk 8 above)
Comment (Sachinbisht27) [reply api/services/user_activities_service.py:59]: Yes, Sure.

File: api/services/user_activities_service.py (hunk 9 above)
Comment (Sachinbisht27) [reply api/services/user_activities_service.py:72]: We can remove them from here as not needed while creating new activity. We want the default value as null only.

**Review by Satendra-SR — APPROVED** [review]

**Pull request merged**